src/node/blockstorage.cpp

```diff
@@ -938,22 +938,31 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
             LogError("FindUndoPos failed");
             return false;
         }
-        // Open history file to append
-        AutoFile fileout{OpenUndoFile(pos)};
-        if (fileout.IsNull()) {
-            LogError("OpenUndoFile failed");
-            return FatalError(m_opts.notifications, state, _("Failed to write undo data."));
-        }
 
-        // Write index header
-        fileout << GetParams().MessageStart() << blockundo_size;
-        pos.nPos += BLOCK_SERIALIZATION_HEADER_SIZE;
         {
-            // Calculate checksum
-            HashWriter hasher{};
-            hasher << block.pprev->GetBlockHash() << blockundo;
-            // Write undo data & checksum
-            fileout << blockundo << hasher.GetHash();
+            // Open history file to append
+            AutoFile fileout{OpenUndoFile(pos)};
+            if (fileout.IsNull()) {
+                LogError("OpenUndoFile failed");
+                return FatalError(m_opts.notifications, state, _("Failed to write undo data."));
+            }
+
+            // Write index header
+            fileout << GetParams().MessageStart() << blockundo_size;
+            pos.nPos += BLOCK_SERIALIZATION_HEADER_SIZE;
+            {
+                // Calculate checksum
+                HashWriter hasher{};
+                hasher << block.pprev->GetBlockHash() << blockundo;
+                // Write undo data & checksum
+                fileout << blockundo << hasher.GetHash();
+            }
+
+            // Make sure `AutoFile` goes out of scope before we call `FlushUndoFile`
+            if (fileout.fclose()) {
```

Comment:
In commit 6640dd52c9fcb85d77f081780c02ee37b8089091

Introducing this `if` statement seems to be ephemeral, but even so, I'm not sure if its effects are really as intended. Previously `fclose` failing was just ignored. Is it really safe to add this early return now instead of just letting the `Autofile` destructor do its thing? Maybe I am missing something, but this will have effects on when `ConnectBlock` might fail, and I'm not sure if that is intended. There is also no similar check when we are writing blocks.

I saw the discussion in https://github.com/bitcoin/bitcoin/pull/31551#discussion_r2031325418, and https://github.com/bitcoin/bitcoin/pull/31551#discussion_r2036694308 , but I still don't quite follow why a failed close is dangerous here, and not elsewhere.

Reply:
It's an intermediary commit, the final code contains a flush only: https://github.com/bitcoin/bitcoin/pull/31551/files#diff-114c2880ec1ff2c5293ac65ceda0637bf92c05745b74b58410585a549464a33fR963

Reply:
So why is it added at all?

Reply:
Yes, I read that, and did not understand how this makes it more self-contained, or complete. The commit message also said nothing about adding a new error condition.

Reply:
It's closed now uniformly in https://github.com/bitcoin/bitcoin/pull/29307/files#diff-114c2880ec1ff2c5293ac65ceda0637bf92c05745b74b58410585a549464a33fR965-R972 - does that address your concern?

Reply:
Not really, but I guess we can continue over there.